### PR TITLE
Pcre2 3194 v6

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3194

Describe changes:
- Use PCRE2 instead of pcre for transform pcrexform
- qa : remove some obsolete files
- pcrexform : checks that we have exactly one substring capture in a signature
- transforms : use generic `Free` functions everywhere
- migrate detect-pcre to pcre2

After that, the goal is to remove pcre everywhere else cf `git grep pcre_compile`

suricata-verify-pr: 480

https://github.com/OISF/suricata-verify/pull/480

Replaces https://github.com/OISF/suricata/pull/6026 with migrating detect-pcre to pcre2 also for keyword parsing itself...

### CODE REVIEW

File: src/detect-transform-pcrexform.c [hunk 1]
```diff
@@ -84,14 +91,37 @@ static int DetectTransformPcrexformSetup (DetectEngineCtx *de_ctx, Signature *s,
         SCReturnInt(-1);
     }
 
-    if (!DetectSetupParseRegexesOpts(regexstr, pxd, 0)) {
+    int en;
```
Comment (jlucovsky) [inline src/detect-transform-pcrexform.c:94]: Pre-existing -- can you fix the `SCCalloc`? The params are reversed :-( My bad.

File: src/detect-transform-pcrexform.c [hunk 2]
```diff
@@ -61,9 +65,12 @@ static void DetectTransformPcrexformFree(DetectEngineCtx *de_ctx, void *ptr)
 {
     if (ptr != NULL) {
         DetectTransformPcrexformData *pxd = (DetectTransformPcrexformData *) ptr;
+        pcre2_code_free(pxd->regex);
+        pcre2_match_data_free(pxd->match_data);
```
Comment (jlucovsky) [inline src/detect-transform-pcrexform.c:69]: I think you need a null check here ... this function can be called with `match_data = NULL`.

File: src/detect-pcre.c [hunk 3]
```diff
@@ -71,39 +71,23 @@
...
-            DetectThreadCtxGetKeywordThreadCtx(det_ctx, pd->thread_ctx_jit_stack_id);
-        if (jit_stack) {
-            SCLogDebug("Using jit_stack %p", jit_stack);
-            return pcre_jit_exec(regex->regex, regex->study, str, strlen,
-                                start_offset, options, ovector, ovector_size,
-                                jit_stack);
-        }
-    }
-#endif
     /* Fallback if registration during setup failed */
```
Comment (jlucovsky) [inline src/detect-pcre.c:88]: This comment no longer applies.

File: configure.ac [hunk 4]
```diff
@@ -637,6 +637,35 @@
         exit 1
     fi
 
+    PCRE2=""
+    #AC_CHECK_HEADER(pcre2.h,,[AC_MSG_ERROR(pcre2.h not found ...)])
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply configure.ac:641]: ok

File: src/detect-transform-pcrexform.c (hunk 1 above)
Comment (catenacyber) [reply src/detect-transform-pcrexform.c:94]: What do you mean ?

File: src/detect-transform-pcrexform.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-transform-pcrexform.c:69]: How could it be `NULL` ?

+ This behaves well for NULL cf https://www.pcre.org/current/doc/html/pcre2_match_data_free.html

File: src/detect-pcre.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-pcre.c:88]: removing

**Comment by catenacyber** [comment]
Replaced by https://github.com/OISF/suricata/pull/6036

**Pull request closed without merging**